controllers/ocirepository_controller.go

```diff
@@ -0,0 +1,909 @@
+	MaxConcurrentReconciles   int
+	DependencyRequeueInterval time.Duration
+	RateLimiter               ratelimiter.RateLimiter
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *OCIRepositoryReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return r.SetupWithManagerAndOptions(mgr, OCIRepositoryReconcilerOptions{})
+}
+
+func (r *OCIRepositoryReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, opts OCIRepositoryReconcilerOptions) error {
+	r.requeueDependency = opts.DependencyRequeueInterval
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&sourcev1.OCIRepository{}, builder.WithPredicates(
+			predicate.Or(predicate.GenerationChangedPredicate{}, predicates.ReconcileRequestedPredicate{}),
+		)).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: opts.MaxConcurrentReconciles,
+			RateLimiter:             opts.RateLimiter,
+		}).
+		Complete(r)
+}
+
+// +kubebuilder:rbac:groups=source.toolkit.fluxcd.io,resources=ocirepositories,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=source.toolkit.fluxcd.io,resources=ocirepositories/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=source.toolkit.fluxcd.io,resources=ocirepositories/finalizers,verbs=get;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+
+func (r *OCIRepositoryReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, retErr error) {
+	start := time.Now()
+	log := ctrl.LoggerFrom(ctx).
+		// Sets a reconcile ID to correlate logs from all suboperations.
+		WithValues("reconcileID", uuid.NewUUID())
+
+	// logger will be associated to the new context that is
+	// returned from ctrl.LoggerInto.
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// Fetch the OCIRepository
+	obj := &sourcev1.OCIRepository{}
+	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Record suspended status metric
+	r.RecordSuspend(ctx, obj, obj.Spec.Suspend)
+
+	// Return early if the object is suspended
+	if obj.Spec.Suspend {
+		log.Info("reconciliation is suspended for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// Initialize the patch helper with the current version of the object.
+	patchHelper, err := patch.NewHelper(obj, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// recResult stores the abstracted reconcile result.
+	var recResult sreconcile.Result
+
+	// Always attempt to patch the object and status after each reconciliation
+	// NOTE: The final runtime result and error are set in this block.
+	defer func() {
+		summarizeHelper := summarize.NewHelper(r.EventRecorder, patchHelper)
+		summarizeOpts := []summarize.Option{
+			summarize.WithConditions(ociRepositoryReadyCondition),
+			summarize.WithReconcileResult(recResult),
+			summarize.WithReconcileError(retErr),
+			summarize.WithIgnoreNotFound(),
+			summarize.WithProcessors(
+				summarize.RecordContextualError,
```

Comment:
This is an old result processor which processed contextual error. This works with the contextual error like the `Event` error. But a few months ago, we moved away from contextual error to configurable error and introduced `Generic` error type as part of https://github.com/fluxcd/source-controller/pull/724 . And for handling the new configurable errors, `ErrorActionHandler` was introduced https://github.com/fluxcd/source-controller/commit/30fe0dc6aa0527acf57586ab862eff586b102455.

Maybe a month ago, when I asked to replace the Event error with Generic error in this file, it was my bad that I forgot to mention about the error processor change. So let's change this to use `summarize.ErrorActionHandler`.

I discovered this issue when I created a OCIRepo with a non-existing URL which results in artifact pull fail. We'd expect to receive a notification/event about this failure but nothing was showing up. The reason was that the Generic error returned from `reconcileSource()` is not known to `RecordContextualError` and it wasn't being handled. Replacing with `ErrorActionHandler` fixes it and we get proper event and notification for such failure.

```
9m40s       Warning   OCIArtifactPullFailed        ocirepository/podinfo-oci   failed to pull artifact from 'oci://ghcr.io/stefanprodan/manifests/podinfo/foo': GET https://ghcr.io/v2/stefanprodan/manifests/podinfo/foo/manifests/6.1.7: MANIFEST_UNKNOWN: manifest unknown
```

Reply:
With this, the failure recovery notifications work properly
![oci-recovery](https://user-images.githubusercontent.com/614105/183088566-1ad1ab5f-4fb9-4e4c-88df-12568c8f275c.png)
Failure notification and then recovery on fixing the issue.